File: src/linear_model.ts
export const TRAINING_SET = [
  [1, 2],
  [2, 4],
  [4, 8],
  [10, 20],
];

export function linear_model() {
  const epsilon = 1e-3;
  const k = 1e-3;
  const gen = 500;
  let w = 7.5;

  function cost(w: number) {
    let cost = 0;
    for (let i = 0; i < TRAINING_SET.length; i++) {
      const x: number = TRAINING_SET[i][0];
      const f: number = x * w;
      const expected: number = TRAINING_SET[i][1];
      const distance: number = f - expected;
      cost += distance * distance;
    }
    return cost / TRAINING_SET.length;
  }

  for (let i = 0; i < gen; i++) {
    console.log(cost(w));
    const dCost = (cost(w + epsilon) - cost(w)) / epsilon;
    w -= k * dCost;
    console.log(cost(w));
  }
}

File: zig/src/linear_model.zig
const std = @import("std");
const alg = @import("algebra.zig");

const rg = std.rand.DefaultPrng;

const eps: f64 = 1e-3;
const k: f64 = 1e-3;
const gen = 9_999_999;

const training_set = [4][2]f64{
    [_]f64{ 2, 4 },
    [_]f64{ 3, 6 },
    [_]f64{ 4, 8 },
    [_]f64{ 5, 10 },
};

fn cost(w: f64, b: f64) !f64 {
    var c: f64 = 0;
    for (training_set) |row| {
        var x = row[0];
        var e = row[1];
        var y = x * w + b;
        var diff = e - y;
        c += diff * diff;
    }

    c /= 4;
    return c;
}
pub fn linear_model() ![3]f64 {
    var rn = rg.init(0);
    var w = rn.random().float(f64);
    var b = rn.random().float(f64);
    var i: usize = 0;

    while (i < gen) {
        const cache = try cost(w, b);
        const dw = (try cost(w + eps, b) - cache) / eps;
        const db = (try cost(w, b + eps) - cache) / eps;
        w -= k * dw;
        b -= k * db;

        i += 1;
    }

    return [3]f64{ w, b, try cost(w, b) };
}

pub fn main() !void {
    const res = try linear_model();
    std.debug.print("{any}\n", .{res});
}

test "Linear Model" {
    // const y = try linear_model();
    // for (training_set) |row| {
    //     const res = y[0] * row[0] + y[1];
    //     std.debug.print("\n{} * {} + {} = {} exp {}\n", .{ row[0], y[0], y[1], res, row[1] });
    // }
    //
    var r = alg.sigmoid(3);

    std.debug.print("\n{}\n", .{r});
}
